2단계 - 지하철 노선 관리

thddntjr1234 opened the pull request:
안녕하세요 플람님! 2단계 미션 완료했습니다.

몇가지 질문 사항이 있어 같이 코멘트 남기겠습니다.
1. 특정 컨트롤러에 @ModelAttribute 어노테이션을 사용해야 할 필요가 있어 request dto에 setter를 추가
    - dto에도 가능하면 setter를 사용하는 걸 지양하고 싶은데 그러려면 생성자들을 제거해야 함.
    -  테스트에서 생성자를 통해 요청 객체들을 만드는 상황이라 제거할 수 없는 상황
    -  이런 경우엔 setter를 허용하나요? 어떤 키워드로 공부해야 할지 잘 모르겠습니다.

2. 노선 인수 테스트를 진행 도중 조건에 맞는 생성자를 추가하다 보니 생성자 양이 많아짐
    - 이런 생성자가 많아져도 괜찮은 건가요?
    - 빌더를 사용하면 100% 해결이 될지 궁금합니다.

thddntjr1234 committed:
1단계 - 미션 리뷰 반영

- createStation() 파라미터를 이름 문자열을 받도록 수정
- 복잡한 타입을 반환받는 변수는 var 타입으로 변경
- contains() -> conatinsExactly() 변경

thddntjr1234 committed:
2단계 - 지하철 노선 생성 인수 테스트 작성

thddntjr1234 committed:
2단계 - 지하철 노선 생성 인수 테스트 기능 구현

- RestAssured 공통 클래스 분리
- JPA 연관관계 방식으로 엔티티 구성
- Entitn -> DTO로 변환 생성자 추가

thddntjr1234 committed:
2단계 - 지하철 노선 목록 조회 인수 테스트 작성

thddntjr1234 committed:
2단계 - 지하철 노선 목록 조회 테스트 기능 구현

thddntjr1234 committed:
2단계 - 서비스 내의 createLineResponse()를 엔티티 내의 toResponse()로 리팩토링

thddntjr1234 committed:
2단계 - 지하철 노선 조회 인수 테스트 작성

thddntjr1234 committed:
2단계 - 지하철 노선 조회 인수 테스트 기능 구현

thddntjr1234 committed:
2단계 - 지하철 노선 수정 인수 테스트 작성

thddntjr1234 committed:
2단계 - 지하철 노선 수정 인수 테스트 기능 구현

thddntjr1234 committed:
2단계 - 지하철 노선 삭제 인수 테스트 작성

thddntjr1234 committed:
2단계 - 지하철 노선 삭제 인수 테스트 기능 구현

thddntjr1234 committed:
2단계 - 각 테스트가 격리된 환경에서 실행될 수 있도록 truncate sql 추가

thddntjr1234 committed:
2단계 - 노선 생성 메서드 공통화

Flamme1004K commented on src/main/java/subway/LineController.java:
```diff
@@ -0,0 +1,46 @@
+    public ResponseEntity<LineResponse> deleteLine(@ModelAttribute LineRequest lineRequest) {
+        lineService.deleteLine(lineRequest);
+        return ResponseEntity.ok(null);
+    }
```
> 특정 컨트롤러에 @ModelAttribute 어노테이션을 사용해야 할 필요가 있어 request dto에 setter를 추가

모든 Request에 대해 하나의 dto를 사용하는 것은 지양하면 좋을 것 같아요. API의 요청에 대해서는 모두 각 API에 맞게 요청에 대한 DTO를 만드는 것이 좋아요. 왜냐하면 API마다 요청 값이 다르기 때문에 중복 된 의미의 필드가 여러개가 생길 수도 있습니다. 또한 하나의 DTO에 모든 필드가 공유가 되어진다면, 변경에 대해서도 좋지 않습니다. 이에 대해서는 요청(LineRequest)과 응답(LineResponse)은 응답 구조에 따라서 DTO를 모두 나누어주도록 변경해 보세요.

Flamme1004K commented on src/main/java/subway/LineController.java:
```diff
@@ -0,0 +1,46 @@
+    @PutMapping("/line")
+    public ResponseEntity<LineResponse> updateLine(@RequestBody LineRequest lineRequest) {
+        lineService.modifyLine(lineRequest);
+        return ResponseEntity.ok(null);
```
ResponseEntity.ok가 아닌, API 명세처럼 204 no-content를 보내보도록 해보세요.

Flamme1004K commented on src/main/java/subway/Line.java:
```diff
@@ -0,0 +1,79 @@
+    private String name;
+    @Column(length = 30, nullable = false)
+    private String color;
+    @ManyToOne
```
`@OneToOne` 이 아닌 `@ManyToOne`을 사용한 이유를 알 수 있을까요?

thddntjr1234 replied:
강남역이 2호선과 신분당 노선에 포함된 것처럼 한 역은 여러 노선에 포함될 수 있기 때문에 다대일 관계라고 생각해서 `@ManyToOne`을 사용했습니다!

Flamme1004K replied:
station 입장에서는 강남역 2호선과 신분당 노선에 대해 Line이 여러 개일 수는 있어 보입니다. 하지만, 강남역 2호선 하나로 봤을 때는 station이 하나이지 않을까요!?

Flamme1004K commented on src/main/java/subway/Line.java:
```diff
@@ -0,0 +1,79 @@
+        this.upwardStation = upwardStation;
+        this.downwardStation = downwardStation;
+        this.distance = distance;
+    }
```
현재 요구사항으로써 update는 이름과 색만 변경이 되어집니다. 변경해 볼까요?

Flamme1004K commented on src/main/java/subway/Line.java:
```diff
@@ -0,0 +1,79 @@
+        return distance;
+    }
+
+    public LineResponse toResponse() {
+        return new LineResponse(this);
+    }
```
LineResponse가 여러 개가 생기면 점차 Line에 dto 변경 클래스가 생기겠군요?

차라리, 해당 메서드를 LineResponse의 정적 메서드로 만드는 게 어떨까요?

thddntjr1234 replied:
각 Response dto 객체의 정적 메서드로 만들어 보겠습니다~

Flamme1004K commented on src/main/java/subway/StationService.java:
```diff
@@ -32,6 +32,10 @@ public void deleteStationById(Long id) {
     }
 
+    public Station findByStationId(Long stationId) {
+        return stationRepository.findById(stationId).orElse(null);
```
null을 반환하는 것보다는 존재하지 않는다는 예외가 더 좋아보입니다.

Flamme1004K commented on src/test/java/subway/LineAcceptanceTest.java:
```diff
@@ -0,0 +1,145 @@
+
+@DisplayName("지하철 노선 관련 기능")
+@SpringBootTest(webEnvironment = WebEnvironment.DEFINED_PORT)
+@Sql(scripts = "/line-test.sql", executionPhase = Sql.ExecutionPhase.BEFORE_TEST_METHOD)
```
👍

Flamme1004K commented on src/test/java/subway/LineAcceptanceTest.java:
```diff
@@ -0,0 +1,145 @@
+     */
+    @BeforeEach
+    void initStations() {
+        Arrays.asList("장암역", "석남역", "계양역", "송도달빛축제공원역")
+                .forEach(StationCommonApi::createStation);
```
```suggestion
var 장암역 = StationCommonApi.createStation("장암역");
var 석남역 = StationCommonApi.createStation("석남역");
var 계양역 = StationCommonApi.createStation("계양역");
var 송도달빛축제공원역 = StationCommonApi.createStation("송도달빛축제공원역");
```

역에 대해서 LineAcceptanceTest가 가지고 있으면, 인수테스트 시나리오에서 잘 쓰이지 않을까요?

Flamme1004K reviewed (changes requested):
안녕하세요.

2단계 미션 잘 해주셨습니다. 몇 가지 피드백을 남겼으니 확인 후 다시 리뷰 요청 부탁드립니다.

감사합니다.

Flamme1004K commented:
> 특정 컨트롤러에 @ModelAttribute 어노테이션을 사용해야 할 필요가 있어 request dto에 setter를 추가
> 
> dto에도 가능하면 setter를 사용하는 걸 지양하고 싶은데 그러려면 생성자들을 제거해야 함.
> 테스트에서 생성자를 통해 요청 객체들을 만드는 상황이라 제거할 수 없는 상황
> 이런 경우엔 setter를 허용하나요? 어떤 키워드로 공부해야 할지 잘 모르겠습니다.
> 노선 인수 테스트를 진행 도중 조건에 맞는 생성자를 추가하다 보니 생성자 양이 많아짐
> 
> 이런 생성자가 많아져도 괜찮은 건가요?
> 빌더를 사용하면 100% 해결이 될지 궁금합니다.

에 대한 부분은 일단, Request와 Response를 나누고 한 번 더 이야기해 보시죠~

thddntjr1234 committed:
2단계 - 리뷰 내용 반영

1. 요청별 dto 구분
    - LineRequest -> 생성, 업데이트 요청 dto로 구분
2. Entity.toResponse()를 response dto에서 생성하도록 정적 메서드 추가
3. 미션 요구사항에 맞는 HttpStatus를 응답하도록 수정
4. findByStationId null 대신 예외 발생으로 수정
5. 역 정보를 LineAcceptanceTest가 가지도록 수정

thddntjr1234 commented:
안녕하세요 플람님. 리뷰해 주신 내용 반영해서 커밋했습니다!
수정하던 도중 추가로 궁금한 부분이 생겨서 질문 남겨놓겠습니다.

**[질문사항]**
1. 동일한 응답 구조를 가지는 컨트롤러에 대해서도 response dto를 분리해야 하는지?
    - 현재 생성, 단건 조회는 동일한 데이터 응답 구조를 가짐
     - 관리 포인트가 불필요하게 늘어나는 느낌이라 분리할 이유가 없는 것 같아 질문드립니다.

Flamme1004K commented on src/main/java/subway/LineController.java:
```diff
@@ -0,0 +1,46 @@
+    @PutMapping("/lines/{id}")
+    public ResponseEntity<LineResponse> updateLine(@PathVariable Long id, @RequestBody LineUpdateRequest lineRequest) {
+        lineService.updateLine(id, lineRequest);
+        return ResponseEntity.ok(null);
```
```suggestion
        return ResponseEntity.ok().build();
```

로 가능해 보이네요.

Flamme1004K commented on src/main/java/subway/LineCreateRequest.java:
```diff
@@ -0,0 +1,40 @@
+package subway;
+
+public class LineCreateRequest {
```
피드백 반영 👍👍

Flamme1004K commented on src/test/java/subway/LineAcceptanceTest.java:
```diff
@@ -0,0 +1,150 @@
+     */
+    @BeforeEach
+    void initStations() {
+        장암역 = StationCommonApi.createStation("장암역").as(StationResponse.class);
+        석남역 = StationCommonApi.createStation("석남역").as(StationResponse.class);
+        계양역 = StationCommonApi.createStation("계양역").as(StationResponse.class);
+        송도달빛축제공원역 = StationCommonApi.createStation("송도달빛축제공원역").as(StationResponse.class);
```
잘해주셨네요! 👍👍

Flamme1004K reviewed (approved):
안녕하세요! 

피드백 반영 너무 잘해주셨네요! 머지 하도록 하겠습니다.

다음 미션 화이팅하시길 바랍니다!

Flamme1004K commented:
질문에 대해서 피드백을 드릴 때가 되서 코멘트 추가합니다~ 

> 특정 컨트롤러에 @ModelAttribute 어노테이션을 사용해야 할 필요가 있어 request dto에 setter를 추가
> 
> dto에도 가능하면 setter를 사용하는 걸 지양하고 싶은데 그러려면 생성자들을 제거해야 함.
> 테스트에서 생성자를 통해 요청 객체들을 만드는 상황이라 제거할 수 없는 상황
> 이런 경우엔 setter를 허용하나요? 어떤 키워드로 공부해야 할지 잘 모르겠습니다.
> 노선 인수 테스트를 진행 도중 조건에 맞는 생성자를 추가하다 보니 생성자 양이 많아짐
> 
> 이런 생성자가 많아져도 괜찮은 건가요?
> 빌더를 사용하면 100% 해결이 될지 궁금합니다.

해당 부분은 Request를 나눔으로써 자연스럽게 `@ModelAttribute`를 해소하셨을거에요. 저는 Request에 대해서는 모두 클래스로 나누어서 클래스 내에 필요한 필드들만 넣습니다. 이로써 클래스 내에 필요한 명령들이 자연스럽게 정리가 되며, 필요 없는 필드에 대해서, 관련 없는 필드에 대해서 정리할 수 있다고 봅니다. 이렇게 함으로써 결국에는 요청에 대한 생성자도 많이 줄어드는 부분이 있습니다. 

이렇게 함으로써도 옵셔널한 필드가 많은 경우에 인수테스트나 단위테스트에서는 생성자가 많아지는 케이스가 생길 수 있습니다. 이런 경우에는 자연스러운 부분이라 생각하여 생성자를 많이 만들지만, 4개~5개의 생성자가 만들어지는 부분이 생긴다면 이를 빌더로 만들어 정의하고 있긴 합니다.


> 동일한 응답 구조를 가지는 컨트롤러에 대해서도 response dto를 분리해야 하는지?
> 현재 생성, 단건 조회는 동일한 데이터 응답 구조를 가짐
> 관리 포인트가 불필요하게 늘어나는 느낌이라 분리할 이유가 없는 것 같아 질문드립니다.

응답에 대해서는 공통화 시킬 수도 있고, 분리할 수도 있습니다. 현재 요구사항으로는 생성, 단건 조회에 대해 같은 구조이기 때문에 Response를 같이 가져가도 된다고 생각합니다. 다만, 둘 중에 하나가 틀어지게 된다면 그때는 response를 다르게 가져가야 된다고 저는 생각합니다.
